### ISSUE TITLE
Help: How to configure two tuners in master-slave

### ISSUE THREAD

**Issue by lars18th** [catalinii/minisatip#811]
Hi,

Until now I'm using minisatip to drive one or two _independent_ SAT>IP servers. However, now I need to configure two tuners in a master-slave mode, and I need some help.

That's the configuration:
- 4 orbital positions with one DiSeqc 1.1 swith.
- 1 coax cable.
- 1 passthrought 2x1 derivator.
- 1 SAT>IP server (1 tuner) connected as regular 13-18v with 22KHz (master).
- 1 SAT>IP server (1 tuner) connected without any output current (slave).

Then the slave tuner can only work if the first is running and only consuming the same band in the same satellite position. In this scenario: What's the correct configurition for the minisatip parameteres?

I understand that the minisatip will use every time the master tuner, and only if it's in use and a second request is in the same band/position, then it starts the slave tuner. However, in case of running both tuners, if the session that uses the master tuner closes... what does the minisatip logic? It continues using the master tuner (perhaps with a fake `pids=0`) until the second closes?

Please, help me with this configuration.
Thank you!

**Comment by catalinii** [comment]
In case the stream on the master iadapter is closed, minisatip will keep the frontend enabled with no pid filters to allow the slave to continue.
You can use https://github.com/catalinii/minisatip/blob/master/src/minisatip.c#L452 as an example

**Comment by lars18th** [comment]
Hi @catalinii ,

> In case the stream on the master iadapter is closed, minisatip will keep the frontend enabled with no pid filters to allow the slave to continue.

OK, so then it will work as expected. However, I suggest to change the keeping of the frontend enabled with no pid filters to something different... when using the `satipc` module. What you want to say with "no pid filters"? This implies `pids=all` or `pids=none`? From my point of view the most safe way is to use `pids=0` without changing the original parameters of the original session in the master tuner. The reason is because if the tuners are SAT>IP servers and we're using the `satipc` module, then is a waste of resources to use `pids=all`; and `pids=none` will generates troubles with some servers.

So, in short: This master-slave mode works with the `satipc` module without troubles?

**Comment by catalinii** [comment]
Generally you need to specifically add the pid filters. In this case no pid filter is added.

**Comment by lars18th** [comment]
> Generally you need to specifically add the pid filters. In this case no pid filter is added.

Then, the master tuner will run with `pids=all` until the slave closes?

**Comment by catalinii** [comment]
Will run with no pid filter (not receive aby data)

**Comment by lars18th** [comment]
> Will run with no pid filter (not receive aby data)

So, then it uses "pids=none"? In this case, it will generate troubles. I suggest to use a "pids=0" instead.

Please, can you point to me where is the code that maintains the session open in the master adapter? I'll try to make a look inside it.

**Comment by lars18th** [comment]
Hi @catalinii ,

I confirm that it doesn't work as expected. Let me to explain... I use this start parameters:
```
minisatip -f -e 1-3 -a 2:0:0 -U 1,3:1,3:1,3:2 -S 3:1 -s 192.168.1.51 -s 192.168.1.52 -s 192.168.1.53
```
SAT>IP servers are: 51-master, 52-independent (only for sat-pos 4), 53-slave (linked to 51).
Futhermore I have one USB tuner as device 0 that is disable (for this reason `-e 1-3`).

In this environment what I obtain is:
- With `-S 3:1` (like in the example): When I open another transponder in the same band I receive and error.
- With `-S 3:0` or `-S 3:3` or without `-S`: I can open a second transponder in the same band, but when the first closes, the LNB goes off because no voltage.

So, what I'm doing wrong?
Please, help me.

**Comment by lars18th** [comment]
Hi @catalinii ,

The trouble is very annoying because:
- When the second transponder is request (from the client), the function `get_free_adapter()` is called _twice_.
- The call to `compare_slave_parameters()` really works, as it return `1` only for the third tuner (the slave with id 2).
- When searching for a free tuner this line is annoying: https://github.com/catalinii/minisatip/blob/845f2798b30d13d526332887bbe5c9f612e69a01/src/adapter.c#L737 as the code arrives here when checking the slave tuner, but it's never selected.

So, I arrived to the conclusion that the current master-slave code is wrong.
Please, can you check it?

**Comment by lars18th** [comment]
Hi @catalinii ,

This dirty hack resolves the problem of selecting slave tuners:
```
diff --git a/src/adapter.c b/src/adapter.c
index e3a2705..b3ab2f2 100644
--- a/src/adapter.c
+++ b/src/adapter.c
@@ -619,6 +619,8 @@ void dump_pids(int aid) {
 int compare_slave_parameters(adapter *ad, transponder *tp) {
     adapter *master = NULL;

+    LOGM("in compare_slave_parameters()");
+
     if (!ad)
         return 0;
     // is not a slave and does not have a slave adapter
@@ -712,6 +714,8 @@ int get_free_adapter(transponder *tp) {
         LOG("get free adapter %d msys %s requested %s", fe, get_delsys(fe),
             get_delsys(msys));

+    dump_adapters();
+
     if (fe >= 0) {
         if (ad && delsys_match(ad, msys)) {
             match = 0;
@@ -746,6 +750,19 @@ int get_free_adapter(transponder *tp) {
         }
     }

+    // No regular tuners available, check for slave tuners
+    for (i = 0; i < MAX_ADAPTERS; i++) {
+        // first free slave adapter that has the same msys
+        if ((ad = get_adapter_nw(i)) && ad->sid_cnt == 0 &&
+            delsys_match(ad, msys) && compare_slave_parameters(ad, tp) &&
+            ad->sources_pos[tp->diseqc])
+        {
+            LOGM("get free adapter found slave adapter %d", i);
+            return i;
+        }
+    }
+
+
 noadapter:
     LOG("no adapter found for src:%d f:%d pol:%d msys:%d", tp->diseqc, tp->freq,
         tp->pol, tp->sys);
```
However, I'm not sure about side-effects. Mainly because *it doens't works as expected with the satipc module*. The problem is very specific: Even if the adapter continues opened after the 30 seconds in the primary minisatip (the one that executes the satipc), the remote SAT>IP server (hardware or any minisatip) closes the session. Then the LNB goes off. So, as I've described before, with the satipc module it's mandatory to maintain the session opened in the master tuner. For example using a simple `pids=0`; and maintaining the reading RTP socket up and the RTSP session running.

Please, can you review all of this?
Thank you!

**Comment by catalinii** [comment]
Hey,was not aware you use slave with satipc. I use it with vu+/gigablue generally.

Can you upload the log of a simple tuning using the slave adapter?

./minisatip -f -l http

**Comment by lars18th** [comment]
Hi @catalinii ,

I prefer you do the test. You only need _two_ gigablues in parallel connected to the same LNB. Configure one as regular, and for the second disable the LNB power. Then configure minisatip in you PC to use booth devices, one master, the other as slave. And then try the test.

At time, almost two bugs require to be fixed related to master-slave configurations:

1. The selection of one slave tuner. The current implementation fails to select it.
2. When using the satipc module, the master tuner is closed. It's mandatory to not close the session until the slave tuner is closed. This can be done replacing the TEARDOWN with a "pids=0".

I hope you want to address these issues.
